### PR TITLE
Implement CRUD for API web project

### LINKED ISSUE #36: Implement CRUD Operations for Category Entity in ASP.NET Core Web API
Create a Controller:
Develop a CategoryController to handle HTTP requests.
Implement the following actions:
GET /api/Category: Retrieve all categories.
GET /api/Category/{id}: Retrieve a specific category by ID.
POST /api/Category: Create a new category.
PUT /api/Category/{id}: Update an existing category.
DELETE /api/Category/{id}: Delete a category by ID.
Service Layer:
Create an ICategoryService interface to define the CRUD operations.
Implement the CategoryService class to handle business logic and interact with the database.
Repository Pattern:
Implement the ICategoryRepository interface and CategoryRepository class to manage data access.
Use the Unit of Work pattern to coordinate the repositories.
Database Context:
Ensure the ApplicationDbContext is correctly configured to include the Categories DbSet.
Apply necessary migrations to update the database schema.
Error Handling:
Handle potential errors such as invalid data, not found entities, and database exceptions.
Ensure meaningful error messages are returned to the client.
Steps to Reproduce:

Set up the ASP.NET Core Web API project.
Implement the controller, service, and repository as described.
Configure the database context and apply migrations.
Test the CRUD operations using tools like Swagger or Postman.
Expected Outcome: The API should support full CRUD operations for the Category entity, allowing clients to create, read, update, and delete categories through HTTP requests.

### PR DESCRIPTION
**Pull request by ehsan-nadernezhad**
Controller Implementation:
Develop CategoryController to handle HTTP requests for category operations.
Actions include:
GET /api/Category: Retrieve all categories.
GET /api/Category/{id}: Retrieve a specific category by ID.
POST /api/Category: Create a new category.
PUT /api/Category/{id}: Update an existing category.
DELETE /api/Category/{id}: Delete a category by ID.
Service Layer:
Define ICategoryService interface for CRUD operations.
Implement CategoryService to handle business logic and database interactions.
Repository Pattern:
Implement ICategoryRepository interface and CategoryRepository class for data access.
Use the Unit of Work pattern to manage repository coordination.
Database Context:
Configure ApplicationDbContext to include Categories DbSet.
Apply necessary migrations to update the database schema.
Error Handling:
Implement error handling for invalid data, not found entities, and database exceptions.
Ensure meaningful error messages are returned to the client.
Objective: To provide full CRUD functionality for the Category entity, enabling clients to manage categories through the API.

### CODE REVIEW

**Comment by mahmood-ghaem** [comment]
Based on your comment, It seems that you fully understood the topic, Well done.✌

File: Uplift.Service/CategorySerivce.cs [hunk 1]
```diff
@@ -1,25 +1,87 @@
...
+using Microsoft.EntityFrameworkCore;
+using Uplift.DataAccess.Data;
 using Uplift.Models;
 
 namespace Uplift.Service
 {
     public class CategoryService : ICategoryService
     {
-        private readonly IUnitOfWork _unitOfWork;
+        private readonly ApplicationDbContext _context;
```
Comment (mahmood-ghaem) [inline Uplift.Service/CategorySerivce.cs:35]: In the service layer, we don't use the context we use unitOfWork instead. It doesn't mean that it is wrong but based on our architecture we use UOW repository pattern.

File: Uplift.Service/CategorySerivce.cs [hunk 2]
```diff
@@ -1,25 +1,87 @@
...
         }
-        public IEnumerable<Category> GetAllCategories()
+
+        public async Task<Category> GetCategoryByIdAsync(int id)
         {
-            return _unitOfWork.Category.GetAll();
+            return await _context.Categories.FindAsync(id);
+        }
+
+        public async Task CreateCategoryAsync(Category category)
```
Comment (mahmood-ghaem) [inline Uplift.Service/CategorySerivce.cs:52]: That is a good implementation that u are using async methods. 👍

File: Uplift.DataAccess/Migrations/20241007092412_RenameCategoryToCategories.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
+
+#nullable disable
+
+namespace Uplift.DataAccess.Migrations
+{
+    /// <inheritdoc />
+    public partial class RenameCategoryToCategories : Migration
```
Comment (mahmood-ghaem) [inline Uplift.DataAccess/Migrations/20241007092412_RenameCategoryToCategories.cs:8]: That is good that u used capital letter for naming the migration. 👍 Also the name defines the reason of migration.

File: Uplift.Api/Controllers/CategoryController.cs [hunk 4]
```diff
@@ -15,18 +15,72 @@ public CategoryController(ICategoryService categoryService)
             _categoryService = categoryService;
         }
 
-        [HttpPost("create-category")]
-        public IActionResult Create([FromBody] Category model)
+        // GET: api/Category
+        [HttpGet]
+        public async Task<ActionResult<IEnumerable<Category>>> GetCategories()
```
Comment (mahmood-ghaem) [inline Uplift.Api/Controllers/CategoryController.cs:20]: For now, it is fine to use Category here, Just I wanna mention in the future we will define ViewModels for each entity and use VMs here.

File: Uplift.Api/Controllers/CategoryController.cs [hunk 5]
```diff
@@ -15,18 +15,72 @@ public CategoryController(ICategoryService categoryService)
...
+            return Ok(category);
+        }
+
+        // POST: api/Category
+        [HttpPost]
+        public async Task<ActionResult<Category>> CreateCategory([FromBody]Category category)
+        {
+            if (!ModelState.IsValid)
+            {
+                return BadRequest(ModelState);
```
Comment (mahmood-ghaem) [inline Uplift.Api/Controllers/CategoryController.cs:44]: That is good u use relevant responses to the client in order to handle the errors.

**Review by mahmood-ghaem — CHANGES_REQUESTED** [review]
Everything was fine and beyond my expectations, good job. Just one point please use UOW.

File: Uplift.Api/Controllers/CategoryController.cs (hunk 5 above)
Comment (ehsan-nadernezhad) [reply Uplift.Api/Controllers/CategoryController.cs:44]: Thanks. I need to read more about this.

File: Uplift.Api/Controllers/CategoryController.cs (hunk 4 above)
Comment (ehsan-nadernezhad) [reply Uplift.Api/Controllers/CategoryController.cs:20]: Thank you. I wrote it down.

File: Uplift.Service/CategorySerivce.cs (hunk 1 above)
Comment (ehsan-nadernezhad) [reply Uplift.Service/CategorySerivce.cs:35]: Thank you for note. I wrote it  down.

File: Uplift.Service/CategorySerivce.cs [hunk 6]
```diff
@@ -1,87 +1,26 @@
...
+        private readonly IUnitOfWork _unitOfWork;
 
-        public async Task<Category> GetCategoryByIdAsync(int id)
+        public CategoryService(IUnitOfWork unitOfWork) 
         {
-            return await _context.Categories.FindAsync(id);
+            _unitOfWork = unitOfWork;
         }
 
-        public async Task CreateCategoryAsync(Category category)
```
Comment (mahmood-ghaem) [inline Uplift.Service/CategorySerivce.cs]: In the future, we will refactor our UOW in order to support async methods.

File: Uplift.DataAccess/Migrations/20241009130403_UpdateCategorySchema.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿using Microsoft.EntityFrameworkCore.Migrations;
```
Comment (mahmood-ghaem) [inline Uplift.DataAccess/Migrations/20241009130403_UpdateCategorySchema.cs:1]: This migration does nothing so u can remove it (by command 😉)

**Review by mahmood-ghaem — APPROVED** [review]
Thanks, everything looks good, I approve it. Please take to your account by the next PR, remove the unnecessary migration.

**Comment by mahmood-ghaem** [comment]
@ehsan-nadernezhad Once all comments are resolved, the merge is not blocked.

**Pull request merged**